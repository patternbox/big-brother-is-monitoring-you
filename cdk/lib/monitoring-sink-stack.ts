import { Construct } from "constructs";
import { CfnOutput, Stack, StackProps, CfnParameter } from "aws-cdk-lib";
import * as oam from 'aws-cdk-lib/aws-oam';

export class MonitoringSinkStack extends Stack {

  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props)

    const linkedAccounts = new CfnParameter(this, 'LinkedAccounts', {
      type: 'CommaDelimitedList',
      description: 'A list of linked monitoring linked accounts',
    }).valueAsList

    const policy = {
      Version: "2012-10-17",
      Statement: [
        {
          Effect: "Allow",
          Resource: "*",
          Action: ["oam:CreateLink", "oam:UpdateLink"],
          Principal: {
            "AWS": linkedAccounts,
          },
          Condition: {
            "ForAllValues:StringEquals": {
              "oam:ResourceTypes": [
                "AWS::CloudWatch::Metric",
                "AWS::Logs::LogGroup",
                "AWS::XRay::Trace"
              ]
            }
          },
        }
      ]
    }

    const cfnSink = new oam.CfnSink(this, 'MonitoringSink', {
      name: this.stackName,
      policy: policy,
    })

    new CfnOutput(this, 'MonitoringSinkArn', {
      value: cfnSink.attrArn,
      exportName: this.stackName + ':MonitoringSinkArn',
    })
  }
}
