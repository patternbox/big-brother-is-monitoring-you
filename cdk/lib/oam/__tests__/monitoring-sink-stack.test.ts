import { App } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { MonitoringSinkStack } from '../monitoring-sink-stack'

describe('MonitoringSinkStack', () => {

  it('should match snapshot', () => {
    const app = new App()

    const stack = new MonitoringSinkStack(app, 'MonitoringSinkStack', {})

    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot()
  })

})
