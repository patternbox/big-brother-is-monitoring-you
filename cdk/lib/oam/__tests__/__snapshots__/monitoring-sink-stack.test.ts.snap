// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitoringSinkStack should match snapshot 1`] = `
Object {
  "Outputs": Object {
    "MonitoringSinkArn": Object {
      "Export": Object {
        "Name": "MonitoringSinkStack:MonitoringSinkArn",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "MonitoringSink",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LinkedAccounts": Object {
      "Description": "A list of linked monitoring linked accounts",
      "Type": "CommaDelimitedList",
    },
  },
  "Resources": Object {
    "MonitoringSink": Object {
      "Properties": Object {
        "Name": "MonitoringSinkStack",
        "Policy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "oam:CreateLink",
                "oam:UpdateLink",
              ],
              "Condition": Object {
                "ForAllValues:StringEquals": Object {
                  "oam:ResourceTypes": Array [
                    "AWS::CloudWatch::Metric",
                    "AWS::Logs::LogGroup",
                    "AWS::XRay::Trace",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Ref": "LinkedAccounts",
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::Oam::Sink",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
